docs/src/ffi-error-handling.md

```diff
@@ -0,0 +1,249 @@
+
+Rust, on the other hand, aborts the current process whenever `panic!()` is called.  This is, quite clearly, incompatible
+with Postgres' error handling approach.  
+
+It's technically incompatible, it's spiritually incompatible, and it robs Postgres of the opportunity to cleanly rollback 
+the current transaction (Postgres is supposed to be tolerant of such situations, but who wants to test Postgres' 
+recoverability in production?).
+
+Conversely, Postgres' `sigsetjmp/siglongjmp` approach is as egregiously incompatible with Rust.  `siglongjmp` will blindly
+jump over Rust stack frames, leaking Rust-allocated memory, ignoring `trait Drop` implementations, and denying Rust code 
+any opportunity to participate in error handling.
+
+
+# A Wolf, a goat, and some cabbage
+
+pgrx uses two different approaches to protect these FFI boundaries.  While both are implemented in Rust, one protects
+Rust from Postgres `setlongjmp` ERRORs and the other protects Postgres from Rust `panic()!`s.  To make things confusing
+they're both called `#[pg_guard]`.
+
+Essentially, pgrx needs to guard two styles of `extern "C"` functions.  One style is the `extern "C" {}` block that 
+declares a function lives "somewhere else" (in our case, the Postgres process in which the pgrx extension is loaded).
+The other style is `extern "C" fn foo() { ... }` functions that are written in Rust and might be passed to Postgres (for
+it to later call) via a standard function pointer.
+
+## Guarding Postgres Internal Functions
+
+pgrx uses the `bindgen` tool to generate "bindings" for exported Postgres symbols.  Postgres' source header files (`*.h`)
+are read, parsed, and transformed, as much as bindgen knows how, into Rust declarations.  In the case of exported functions,
+bindgen generates blocks similar to:
+
+```rust
+extern "C" {
+    pub fn palloc(size: Size) -> *mut ::std::os::raw::c_void;
+    // ... many more internal Postgres function definitions here ...
+}
+```
+
+Then, pgrx' `build.rs` process rewrites these functions into something similar to this:
+
+```rust
+#[pg_guard]
+extern "C" {
+    pub fn palloc(size: Size) -> *mut ::std::os::raw::c_void;
+    // ... many more internal Postgres function definitions here ...
+}
+```
+
+This form of the `#[pg_guard]` macro then walks the `extern "C" {}` block items and writes new function declarations for 
+each. This expansion looks similar to:
+
+```rust
+pub extern "C" fn palloc(size: Size) -> *mut ::std::os::raw::c_void {
+    extern "C" {
+        pub fn palloc(size: Size) -> *mut ::std::os::raw::c_void;
+        // ... many other function definitions here ...
+    }
+    
+    unsafe {
+        crate::ffi::pg_guard_ffi_boundary(|| palloc(size))
+    }
+}
+```
+
+Essentially, in this usage, `#[pg_guard]` generates standalone wrapper functions that delegate to pgrx' `pg_guard_ffi_boundary(|| ...)`
+function.  This function sets up pgrx' own `sigsetjmp` point, lies to Postgres' exception handling stack about where it's 
```

Comment:
ehh.  I'll expound on that, I guess.  I mentally consider it a lie, because **we** are updating Postgres' exception stack pointer thingiemabob to point to **our** code.  Postgres doesn't expect to jump into some Rust stack frame.  Fortunately, it doesn't need to know.